### PR TITLE
Skip license processing for top-level source directories that are unchanged

### PR DESCRIPTION
**Pull request by jason-simmons**
See https://github.com/flutter/flutter/issues/8106

### CODE REVIEW

**Comment by jason-simmons** [comment]
@Hixie

**Comment by eseidelGoogle** [comment]
Curious how big a win this is? :)

**Comment by jason-simmons** [comment]
For typical changes that only affect files in the flutter/ tree and do not change dart/third_party/etc,
this can sidestep ~97% of the processing work.

File: tools/licenses/lib/main.dart [hunk 1]
```diff
@@ -2237,53 +2271,105 @@ class Progress {
...
 
-void main(List<String> arguments) {
-  if (arguments.length != 1) {
-    print('Usage: dart lib/main.dart path/to/engine/root/src');
+Future<Null> main(List<String> arguments) async {
+  final ArgParser parser = new ArgParser()
+    ..addOption('src', help: 'The root of the engine source')
+    ..addOption('out', help: 'The directory where output is written')
+    ..addOption('golden', help: 'The directory containing golden results')
+    ..addFlag('release', help: 'Print output in the format used for product releases');
```
Comment (chinmaygarde) [inline tools/licenses/lib/main.dart]: Thanks for these flags. It makes the usage so much clearer.

File: tools/licenses/lib/main.dart [hunk 2]
```diff
@@ -1235,6 +1242,33 @@ class RepositoryDirectory extends RepositoryEntry implements LicenseSource {
...
+        yield file;
+    }
+    for (RepositoryDirectory directory in _subdirectories) {
+      yield* directory._allFiles;
+    }
+  }
+
+  Stream<List<int>> _signatureStream(List files) async* {
+    for (RepositoryLicensedFile file in files) {
+      yield file.io.fullName.codeUnits;
```
Comment (chinmaygarde) [inline tools/licenses/lib/main.dart]: Is this really necessary? I am slightly worried about case sensitivity of the file system returning different results per platform. You're comparing the file contents anyway which should catch all cases.

Comment (jason-simmons) [inline tools/licenses/lib/main.dart]: This is intended to catch renaming of files (which will affect the file paths listed in the output)

Comment (chinmaygarde) [inline tools/licenses/lib/main.dart]: Oh. Good point. I forget the file names were also present in that list.

**Comment by Hixie** [comment]
Travis didn't run it?

The code LGTM but I'd like to see Travis run it first. :-)

**Comment by jason-simmons** [comment]
Travis is passing now.

Also modified the script to always run the full license check on the flutter tree.  The goal is to avoid requiring updates to the golden signature on flutter code changes that don't affect the license output.

**Pull request merged**